log4j-api/src/main/java/org/apache/logging/log4j/spi/internal/DefaultScopedContextProvider.java

```diff
@@ -0,0 +1,421 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.spi.internal;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.ScopedContext;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.spi.ScopedContextProvider;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.StringMap;
+
+/**
+ * An implementation of {@link ScopedContextProvider} that uses the simplest implementation.
+ * @since 2.24.0
+ */
+public class DefaultScopedContextProvider implements ScopedContextProvider {
+
+    public static final Logger LOGGER = StatusLogger.getLogger();
+    public static final ScopedContextProvider INSTANCE = new DefaultScopedContextProvider();
+
+    private final ThreadLocal<Instance> scopedContext = new ThreadLocal<>();
+
+    /**
+     * Returns an immutable Map containing all the key/value pairs as Object objects.
+     * @return The current context Instance.
+     */
+    private Optional<Instance> getContext() {
+        return Optional.ofNullable(scopedContext.get());
+    }
+
+    /**
+     * Add the ScopeContext.
+     * @param context The ScopeContext.
+     */
+    private void addScopedContext(final Instance context) {
+        Instance current = scopedContext.get();
+        scopedContext.set(context);
+    }
+
+    /**
+     * Remove the top ScopeContext.
+     */
+    private void removeScopedContext() {
+        final Instance current = scopedContext.get();
+        if (current != null) {
+            if (current.parent != null) {
+                scopedContext.set(current.parent);
+            } else {
+                scopedContext.remove();
+            }
+        }
+    }
+
+    @Override
+    public Map<String, ?> getContextMap() {
+        final Optional<Instance> context = getContext();
+        return context.isPresent()
+                        && context.get().contextMap != null
+                        && !context.get().contextMap.isEmpty()
+                ? Collections.unmodifiableMap(context.get().contextMap)
+                : Collections.emptyMap();
```

Comment:
Due to `setupContext`, the `contextMap` field should be not null here.

As for the immutability of the map, we could ensure it in the constructor of `Instance`. This code could be reduced to:
```java
context.map(c -> c.contextMap).orElse(Collections.emptyMap());
```

Reply:
With the changes this is no longer applicable.